code/datums/helper_datums/teleport.dm

```diff
@@ -198,36 +198,27 @@
 		var/z = pick(zlevels)
 		var/random_location = locate(x,y,z)
 
-		if(!isfloorturf(random_location))
-			continue
-		var/turf/simulated/floor/F = random_location
-		if(!F.air)
-			continue
-
-		var/datum/gas_mixture/A = F.air
-
-		// Can most things breathe?
-		if(A.sleeping_agent)
-			continue
-		if(A.oxygen < 16)
-			continue
-		if(A.toxins)
-			continue
-		if(A.carbon_dioxide >= 10)
-			continue
-
-		// Aim for goldilocks temperatures and pressure
-		if((A.temperature <= 270) || (A.temperature >= 360))
-			continue
-		var/pressure = A.return_pressure()
-		if((pressure <= 20) || (pressure >= 550))
-			continue
-
-		if(extended_safety_checks)
-			if(islava(F)) //chasms aren't /floor, and so are pre-filtered
-				var/turf/simulated/floor/plating/lava/L = F
-				if(!L.is_safe())
-					continue
-
+		if(safety_turf(random_location))
 		// DING! You have passed the gauntlet, and are "probably" safe.
-		return F
+			var/turf/simulated/floor/F = random_location
+			return F
+
+// Checks if the turf is safe to be on
+/proc/safety_turf(location)
```

Comment:
Could you define this proc on `/turf` level instead? And have the turf types override the implementation?
Saves quite a lot of checks and makes it more OO.
The implementation on `/turf` can just return `FALSE` and really only floor turfs and lava turfs need to override it.
Also makes the next question easier, can you also refactor it to use boolean values `TRUE/FALSE` and use early returns etc?

Danke for taking the extra step by the way